ddcommon-ffi/src/array_queue.rs

```diff
@@ -0,0 +1,294 @@
+// Copyright 2021-Present Datadog, Inc. https://www.datadoghq.com/
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::Error;
+use anyhow::Context;
+use std::ffi::c_void;
+
+#[repr(C)]
+// A simple wrapper around crossbeam_queue::ArrayQueue<*mut c_void>, which is a lock free
+// bounded multi-producer and multi-consumer (MPMC) queue.
+pub struct ArrayQueue {
+    // The actual type here should be *mut crossbeam_queue::ArrayQueue<*mut c_void>.
+    // However, cbindgen does not use the module name crossbeam_queue to generate the C header.
+    // So we use *mut c_void here and cast it to the correct type in the FFI functions.
+    inner: *mut c_void,
```

Comment:
I wonder if we can just make this a `Box<ArrayQueue>` https://github.com/rust-lang/unsafe-code-guidelines/issues/157

Reply:
Changed to use `NonNull<ArrayQueue>`, thanks for pointing this out. And added a comment why `NonNull` instead of `Box`